pkg/hive/job/job.go

```diff
@@ -0,0 +1,479 @@
+// different kinds of logic. No matter the job type, the function provided to is always called with a context which
+// is bound to the lifecycle of the cell.
+type Group interface {
+	Add(...Job)
+	hive.HookInterface
+}
+
+// Job in an interface that describes a unit of work which can be added to a Group. This interface contains unexported
+// methods and thus can only be implemented by functions in this package such as OneShot, Timer, or Observer.
+type Job interface {
+	start(ctx context.Context, wg *sync.WaitGroup, options options) error
+}
+
+type group struct {
+	options options
+
+	wg *sync.WaitGroup
+
+	mu         lock.Mutex
+	ctx        context.Context
+	cancel     context.CancelFunc
+	queuedJobs []Job
+}
+
+type options struct {
+	pprofLabels pprof.LabelSet
+	logger      logrus.FieldLogger
+	shutdowner  hive.Shutdowner
+}
+
+type groupOpt func(o *options)
+
+// WithLogger replaces the default logger with the given logger, useful if you want to add certain fields to the logs
+// created by the group/jobs.
+func WithLogger(logger logrus.FieldLogger) groupOpt {
+	return func(o *options) {
+		o.logger = logger
+	}
+}
+
+// WithPprofLabels adds pprof labels which will be added to the goroutines spawned for the jobs and thus included in
+// the pprof profiles.
+func WithPprofLabels(pprofLabels pprof.LabelSet) groupOpt {
+	return func(o *options) {
+		o.pprofLabels = pprofLabels
+	}
+}
+
+var _ hive.HookInterface = (*group)(nil)
+
+// Start implements the hive.HookInterface interface
+func (jg *group) Start(_ hive.HookContext) error {
+	jg.mu.Lock()
+	defer jg.mu.Unlock()
+
+	jg.ctx, jg.cancel = context.WithCancel(context.Background())
+
+	jg.wg.Add(len(jg.queuedJobs))
+	for _, job := range jg.queuedJobs {
+		pprof.Do(jg.ctx, jg.options.pprofLabels, func(ctx context.Context) {
+			go jg.startJob(job, ctx)
+		})
+	}
+	// Nil the queue once we start so it can be GC'ed
+	jg.queuedJobs = nil
+
+	return nil
+}
+
+// Stop implements the hive.HookInterface interface
+func (jg *group) Stop(stopCtx hive.HookContext) error {
+	jg.mu.Lock()
+	defer jg.mu.Unlock()
+
+	done := make(chan struct{})
+	go func() {
+		jg.wg.Wait()
+		close(done)
+	}()
+
+	jg.cancel()
+
+	select {
+	case <-stopCtx.Done():
+		jg.options.logger.Error("Stop hook context expired before job group was done")
+	case <-done:
+	}
+
+	return nil
+}
+
+func (jg *group) Add(jobs ...Job) {
+	jg.mu.Lock()
+	defer jg.mu.Unlock()
+
+	for _, j := range jobs {
+		// The context is only set once the group has been started. If we have not yet started, queue the job.
+		if jg.ctx == nil {
+			jg.queuedJobs = append(jg.queuedJobs, j)
+			return
+		}
+
+		jg.wg.Add(1)
+		pprof.Do(jg.ctx, jg.options.pprofLabels, func(ctx context.Context) {
+			go jg.startJob(j, ctx)
+		})
+	}
+}
+
+func (jg *group) startJob(job Job, ctx context.Context) {
+	err := job.start(ctx, jg.wg, jg.options)
+	if err != nil {
+		jg.options.logger.WithError(err).Error("Job start returned an error")
```

Comment:
Do we need to log that error here. It seems it's already logged in all `.start` methods. I'd say that logging once either here on in `.start` is probably enough.

We instead also could bubble up the error to the callers, combine them there and return them to their respective callers instead of logging them in this package.

Reply:
We can't report the errors back to the caller since the actual job start is performed in a goroutine and might be queued. I have removed this central logging point and ensured that the `job.Start` implementations all log any errors internally.

I have also added a section to the "Guide to Hive" docs.

Reply:
:+1: thanks for clarifying.